core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -0,0 +1,198 @@
+                 const ReducerType& reducer)
+      : m_functor(f),
+        m_policy(p),
+        m_reducer(reducer),
+        m_result_ptr(reducer.view().data()) {
+    // FIXME_SYCL custom reducer not yet implemented
+    if (m_result_ptr == nullptr) std::abort();
+  }
+
+ private:
+  template <typename TagType>
+  std::enable_if_t<std::is_void<TagType>::value> exec(reference_type update) {
+    using member_type = typename Policy::member_type;
+    member_type e     = m_policy.end();
+    for (member_type i = m_policy.begin(); i < e; ++i) m_functor(i, update);
+  }
+
+  template <typename TagType>
+  std::enable_if_t<!std::is_void<TagType>::value> exec(reference_type update) {
+    using member_type = typename Policy::member_type;
+    member_type e     = m_policy.end();
+    for (member_type i = m_policy.begin(); i < e; ++i)
+      m_functor(TagType{}, i, update);
+  }
+
+  template <typename PolicyType, typename Functor>
+  void sycl_direct_launch(const PolicyType& policy,
+                          const Functor& functor) const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    cl::sycl::queue& q = *instance.m_queue;
+
+    q.submit([this, functor, policy](cl::sycl::handler& cgh) {
+      // FIXME_SYCL a local size larger than 1 doesn't work for all cases
+      cl::sycl::nd_range<1> range(policy.end() - policy.begin(), 1);
+
+      value_type identity = 0;
+      ValueInit::init(functor, m_result_ptr);
+      auto reduction =
+          cl::sycl::intel::reduction(m_result_ptr, identity, std::plus<>());
+
+      cgh.parallel_for(
+          range, reduction, [=](cl::sycl::nd_item<1> item, auto& sum) {
+            const typename Policy::index_type id = item.get_global_id(0);
+            value_type partial                   = identity;
+            if constexpr (std::is_same<WorkTag, void>::value)
+              functor(id, partial);
+            else
+              functor(WorkTag(), id, partial);
+            sum.combine(partial);
+          });
+    });
+
+    q.wait();
+  }
+
+  template <typename Functor>
+  void sycl_indirect_launch(const Functor& functor) const {
+    // Convenience references
+    const Kokkos::Experimental::SYCL& space = m_policy.space();
+    Kokkos::Experimental::Impl::SYCLInternal& instance =
+        *space.impl_internal_space_instance();
+    Kokkos::Experimental::Impl::SYCLInternal::IndirectKernelMemory& kernelMem =
+        *instance.m_indirectKernel;
+
+    // Allocate USM shared memory for the functor
+    kernelMem.resize(std::max(kernelMem.size(), sizeof(functor)));
+
+    // Placement new a copy of functor into USM shared memory
+    //
+    // Store it in a unique_ptr to call its destructor on scope exit
+    std::unique_ptr<ReducerTypeFwd, Kokkos::Impl::destruct_delete>
+        kernelFunctorPtr(new (kernelMem.data()) ReducerTypeFwd(functor));
```

Comment:
Did you consider abstracting this away in `SYCLInternal`?

Reply:
I tried to touch all the launch mechanism as little as possible since we want to modify it in a separate pull request anyway, likely incorporating [`c4b0f6b` (#3231)](https://github.com/kokkos/kokkos/pull/3231/commits/c4b0f6ba09f1cda2a9c169e2a3a81c778cd260e1), [`0b0274a` (#3231)](https://github.com/kokkos/kokkos/pull/3231/commits/0b0274a27f617544028633c91179f562090b62ed), [`5708d9d` (#3231)](https://github.com/kokkos/kokkos/pull/3231/commits/5708d9dba70b9527c5db0043cc1c2b7234af24a5), [`f41cb3f` (#3231)](https://github.com/kokkos/kokkos/pull/3231/commits/f41cb3f7dbf1fbe5b8daabd8513cf3a0fd10abc1), [`0427232` (#3231)](https://github.com/kokkos/kokkos/pull/3231/commits/042723264c40aa42d403922614468bb255436164) and [`c08e5b2` (#3231)](https://github.com/kokkos/kokkos/pull/3231/commits/c08e5b225bd37595672948e0d6c9d9b2b5704d77).

Reply:
We agree this is resolved later.